marlowe-playground-client/src/Simulation.purs

```diff
@@ -168,173 +166,109 @@ onEmpty alt [] = alt
-        , Array.fromFoldable (map (inputComposerOracle isEnabled) oracles)
-        ]
+inputComposer :: forall p. Boolean -> Map PubKey (Array ActionInput) -> Array (HTML p Query)
+inputComposer isEnabled actionInputs =
+  onEmpty 
+    [text "No valid inputs can be added to the transaction"]
+    $ actionsForPeople actionInputs
   where
-  ik = Set.fromFoldable (Map.keys inputs)
-
-  cdk = Set.fromFoldable (Map.keys choiceData)
+  kvs :: forall k v. Map k v -> Array (Tuple k v)
+  kvs = Map.toUnfoldable
 
-  people = Set.toUnfoldable (Set.union ik cdk) :: List Person
-
-  oracles = Map.toUnfoldable oracleData :: List (Tuple IdOracle OracleEntry)
+  actionsForPeople :: forall q. Map PubKey (Array ActionInput) -> Array (HTML q Query)
+  actionsForPeople m = foldMap (\(Tuple k v) -> inputComposerPerson isEnabled k v) (kvs m)
 
 inputComposerPerson ::
   forall p.
   Boolean ->
-  Person ->
-  Map Person (List DetachedPrimitiveWIA) ->
-  Map Person (Map BigInteger Choice) ->
+  PubKey ->
+  Array ActionInput ->
   Array (HTML p Query)
-inputComposerPerson isEnabled person inputs choices =
-  Array.concat
-    [ [ h3_
+inputComposerPerson isEnabled person actionInputs =
+      [ h3_
           [ text ("Person " <> show person)
           ]
-      ]
-    , case Map.lookup person inputs of
-        Nothing -> []
-        Just x -> do
-          y <- Array.sortWith (idActionFromDWAI) (Array.fromFoldable x)
-          case y of
-            DWAICommit idAction idCommit val tim -> pure (inputCommit isEnabled person idAction idCommit val tim)
-            DWAIPay idAction idCommit val -> pure (inputPay isEnabled person idAction idCommit val)
-    , case Map.lookup person choices of
-        Nothing -> []
-        Just x -> do
-          (Tuple idChoice choice) <- Map.toUnfoldable x
-          pure (inputChoice isEnabled person idChoice choice)
-    ]
+      ] <> catMaybes (mapWithIndex inputForAction actionInputs)
   where
-  idActionFromDWAI (DWAICommit idAction _ _ _) = idAction
-
-  idActionFromDWAI (DWAIPay idAction _ _) = idAction
+    inputForAction :: Int -> ActionInput -> Maybe (HTML p Query)
+    inputForAction index (DepositInput accountId party value) = Just $ inputDeposit isEnabled person index accountId party value
+    inputForAction index (ChoiceInput choiceId bounds chosenNum) = Just $ inputChoice isEnabled person index choiceId chosenNum bounds
+    inputForAction index (NotifyInput observation) = Just $ inputNotify isEnabled person index observation
 
-inputCommit ::
+inputDeposit ::
   forall p.
   Boolean ->
-  Person ->
-  IdAction ->
-  IdCommit ->
+  PubKey ->
+  Int ->
+  AccountId ->
+  Party ->
   BigInteger ->
-  Timeout ->
   HTML p Query
-inputCommit isEnabled person idAction idCommit val tim =
-  flexRow_
+inputDeposit isEnabled person index accountId party value =
+  let money = wrap value in
+  flexRow_ $
     [ button
         [ class_ $ ClassName "composer-add-button"
         , enabled isEnabled
```

Comment:
interesting, it works well if you use the cursor up and down, I'll see if I can get it to do the same with onKeyPress or whatever.

Reply:
ok, not so easy to deal with the javascript directly so I've changed it so that the button is never disabled but simply does nothing if pressed when the bounds aren't valid (and the redraw causes the error to be displayed).